haystack/templates/source.py

```diff
@@ -0,0 +1,125 @@
+import re
+from enum import Enum
+from pathlib import Path
+from typing import Union, List
+
+import requests
+
+template_file_extension = ".yaml.jinja2"
+
+
+class PredefinedTemplate(Enum):
+    """
+    Enumeration of predefined pipeline templates that can be used to create a `PipelineTemplate` using `TemplateSource`.
+    See TemplateSource.from_predefined for usage.
+    """
+
+    # maintain 1-to-1 mapping between the enum name and the template file name in templates directory
+    QA = "qa"
+    RAG = "rag"
+    INDEXING = "indexing"
+
+    @classmethod
+    def list_all(cls) -> List["PredefinedTemplate"]:
+        """
+        Lists all available pipeline templates by scanning the directory for files with the template file extension.
+
+        :return: A list of strings representing the names of available pipeline templates, excluding the file extension.
+        """
+        directory = Path(__file__).resolve().parent
+        jinja_files = [f for f in directory.iterdir() if f.is_file() and f.name.endswith(template_file_extension)]
+        correct_template_names = [f.name.rsplit(template_file_extension, 1)[0] for f in jinja_files]
+
+        # Map to enums
+        enum_instances = [cls[name.upper()] for name in correct_template_names if name.upper() in cls.__members__]
+        return enum_instances
```

Comment:
Why would we need this function? The idea of having this enum is that every variant is guaranteed to be present, right? Under what circumstances would any of the above be missing its respective template on disk?